README.md
# CertVerify
![made-with-python][made-with-python]
![Python Versions][pyversion-button]
[![Hits](https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2Fpassword123456%2FCertVerify&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=hits&edge_flat=false)](https://hits.seeyoufarm.com)


[pyversion-button]: https://img.shields.io/pypi/pyversions/Markdown.svg
[made-with-python]: https://img.shields.io/badge/Made%20with-Python-1f425f.svg


The CertVerify is a tool designed to detect executable files (exe, dll, sys) that have been signed with untrusted or leaked code signing certificates. The purpose of this tool is to identify potentially malicious files that have been signed using certificates that have been compromised, stolen, or are not from a trusted source.


## Why is this tool needed?
Executable files signed with compromised or untrusted code signing certificates can be used to distribute malware and other malicious software. Attackers can use these files to bypass security controls and to make their malware appear legitimate to victims. This tool helps to identify these files so that they can be removed or investigated further.

As a continuous project of the previous [malware scanner](https://github.com/password123456/malwarescanner), i have created such a tool. This type of tool is also essential in the event of a security incident response.


## Scope of use and limitations
1. The CertVerify cannot guarantee that all files identified as suspicious are necessarily malicious. It is possible for files to be falsely identified as suspicious, or for malicious files to go undetected by the scanner.

2. The scanner only targets code signing certificates that have been identified as malicious by the public community. This includes certificates extracted by malware analysis tools and services, and other public sources. There are many unverified malware signing certificates, and it is not possible to obtain the entire malware signing certificate the tool can only detect some of them. For additional detection, you have to extract the certificate's serial number and fingerprint information yourself and add it to the signatures.


3. The scope of this tool does not include the extraction of code signing information for special rootkits that have already preempted and operated under the kernel, such as FileLess bootkits, or hidden files hidden by high-end technology. In other words, if you run this tool, it will be executed at the user level. Similar functions at the kernel level are more accurate with antirootkit or EDR. Please keep this in mind and focus on the ideas and principles... To implement the principle that is appropriate for the purpose of this tool, you need to development a driver(sys) and run it into the kernel with NT\SYSTEM privileges.

4. Nevertheless, if you want to run this tool in the event of a Windows system intrusion incident, and your purpose is sys files, boot into safe mode or another boot option that does not load the extra driver(sys) files (load only default system drivers) of the Windows system before running the tool. I think this can be a little more helpful.

5. Alternatively, mount the Windows system disk to the Linux and run the tool in the Linux environment. I think this could yield better results.


## Features
- File inspection based on leaked or untrusted certificate lists.
- Scanning includes subdirectories.
- Ability to define directories to exclude from scanning.
- Supports multiprocessing for faster job execution.
- Whitelisting based on certificate subject (e.g., Microsoft subject certificates are exempt from detection).
- Option to skip inspection of unsigned files for faster scans.
- Easy integration with SIEM systems such as Splunk by attaching scan_logs.
- Easy-to-handle and customizable code and function structure.


## And...
- Please let me know if any changes are required or if additional features are needed.
- If you find this helpful, please consider giving it a **"star"**:star2: to support further improvements.


## v1.0.0
- https://github.com/password123456/CertVerify/blob/main/CHANGES

## Preview
![preview](https://github.com/password123456/CertVerify/blob/main/preview.JPG)

## Scan result_log

```
datetime="2023-03-06 20:17:57",scan_id="87ea3e7b-dedc-4016-a43e-5c83f8d27c6e",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\chrome.exe",signature_hash="sha256",serial_number="0e4418e2dede36dd2974c3443afb5ce5",thumbprint="7d3d117664f121e592ef897973ef9c159150e3d736326e9cd2755f71e0febc0c",subject_name="Google LLC",issuer_name="DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1",file_created_at="2023-03-03 23:20:41",file_modified_at="2022-04-14 06:17:04"
datetime="2023-03-06 20:17:58",scan_id="87ea3e7b-dedc-4016-a43e-5c83f8d27c6e",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\LineLauncher.exe",signature_hash="sha256",serial_number="0d424ae0be3a88ff604021ce1400f0dd",thumbprint="b3109006bc0ad98307915729e04403415c83e3292b614f26964c8d3571ecf5a9",subject_name="DigiCert Timestamp 2021",issuer_name="DigiCert SHA2 Assured ID Timestamping CA",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-03-10 18:00:10"
datetime="2023-03-06 20:17:58",scan_id="87ea3e7b-dedc-4016-a43e-5c83f8d27c6e",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\LineUpdater.exe",signature_hash="sha256",serial_number="0d424ae0be3a88ff604021ce1400f0dd",thumbprint="b3109006bc0ad98307915729e04403415c83e3292b614f26964c8d3571ecf5a9",subject_name="DigiCert Timestamp 2021",issuer_name="DigiCert SHA2 Assured ID Timestamping CA",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-04-06 10:06:28"
datetime="2023-03-06 20:17:59",scan_id="87ea3e7b-dedc-4016-a43e-5c83f8d27c6e",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\TWOD_Launcher.exe",signature_hash="sha256",serial_number="073637b724547cd847acfd28662a5e5b",thumbprint="281734d4592d1291d27190709cb510b07e22c405d5e0d6119b70e73589f98acf",subject_name="DigiCert Trusted G4 RSA4096 SHA256 TimeStamping CA",issuer_name="DigiCert Trusted Root G4",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-04-07 09:14:08"
datetime="2023-03-06 20:18:00",scan_id="87ea3e7b-dedc-4016-a43e-5c83f8d27c6e",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\VBoxSup.sys",signature_hash="sha256",serial_number="2f451139512f34c8c528b90bca471f767b83c836",thumbprint="3aa166713331d894f240f0931955f123873659053c172c4b22facd5335a81346",subject_name="VirtualBox for Legacy Windows Only Timestamp Kludge 2014",issuer_name="VirtualBox for Legacy Windows Only Timestamp CA",file_created_at="2023-03-03 23:20:43",file_modified_at="2022-10-11 08:11:56"
datetime="2023-03-06 20:31:59",scan_id="f71277c5-ed4a-4243-8070-7e0e56b0e656",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\chrome.exe",signature_hash="sha256",serial_number="0e4418e2dede36dd2974c3443afb5ce5",thumbprint="7d3d117664f121e592ef897973ef9c159150e3d736326e9cd2755f71e0febc0c",subject_name="Google LLC",issuer_name="DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1",file_created_at="2023-03-03 23:20:41",file_modified_at="2022-04-14 06:17:04"
datetime="2023-03-06 20:32:00",scan_id="f71277c5-ed4a-4243-8070-7e0e56b0e656",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\LineLauncher.exe",signature_hash="sha256",serial_number="0d424ae0be3a88ff604021ce1400f0dd",thumbprint="b3109006bc0ad98307915729e04403415c83e3292b614f26964c8d3571ecf5a9",subject_name="DigiCert Timestamp 2021",issuer_name="DigiCert SHA2 Assured ID Timestamping CA",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-03-10 18:00:10"
datetime="2023-03-06 20:32:00",scan_id="f71277c5-ed4a-4243-8070-7e0e56b0e656",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\LineUpdater.exe",signature_hash="sha256",serial_number="0d424ae0be3a88ff604021ce1400f0dd",thumbprint="b3109006bc0ad98307915729e04403415c83e3292b614f26964c8d3571ecf5a9",subject_name="DigiCert Timestamp 2021",issuer_name="DigiCert SHA2 Assured ID Timestamping CA",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-04-06 10:06:28"
datetime="2023-03-06 20:32:01",scan_id="f71277c5-ed4a-4243-8070-7e0e56b0e656",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\TWOD_Launcher.exe",signature_hash="sha256",serial_number="073637b724547cd847acfd28662a5e5b",thumbprint="281734d4592d1291d27190709cb510b07e22c405d5e0d6119b70e73589f98acf",subject_name="DigiCert Trusted G4 RSA4096 SHA256 TimeStamping CA",issuer_name="DigiCert Trusted Root G4",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-04-07 09:14:08"
datetime="2023-03-06 20:32:02",scan_id="f71277c5-ed4a-4243-8070-7e0e56b0e656",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\VBoxSup.sys",signature_hash="sha256",serial_number="2f451139512f34c8c528b90bca471f767b83c836",thumbprint="3aa166713331d894f240f0931955f123873659053c172c4b22facd5335a81346",subject_name="VirtualBox for Legacy Windows Only Timestamp Kludge 2014",issuer_name="VirtualBox for Legacy Windows Only Timestamp CA",file_created_at="2023-03-03 23:20:43",file_modified_at="2022-10-11 08:11:56"
datetime="2023-03-06 20:33:45",scan_id="033976ae-46cb-4c2e-a357-734353f7e09a",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\chrome.exe",signature_hash="sha256",serial_number="0e4418e2dede36dd2974c3443afb5ce5",thumbprint="7d3d117664f121e592ef897973ef9c159150e3d736326e9cd2755f71e0febc0c",subject_name="Google LLC",issuer_name="DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1",file_created_at="2023-03-03 23:20:41",file_modified_at="2022-04-14 06:17:04"
datetime="2023-03-06 20:33:45",scan_id="033976ae-46cb-4c2e-a357-734353f7e09a",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\LineLauncher.exe",signature_hash="sha256",serial_number="0d424ae0be3a88ff604021ce1400f0dd",thumbprint="b3109006bc0ad98307915729e04403415c83e3292b614f26964c8d3571ecf5a9",subject_name="DigiCert Timestamp 2021",issuer_name="DigiCert SHA2 Assured ID Timestamping CA",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-03-10 18:00:10"
datetime="2023-03-06 20:33:45",scan_id="033976ae-46cb-4c2e-a357-734353f7e09a",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\LineUpdater.exe",signature_hash="sha256",serial_number="0d424ae0be3a88ff604021ce1400f0dd",thumbprint="b3109006bc0ad98307915729e04403415c83e3292b614f26964c8d3571ecf5a9",subject_name="DigiCert Timestamp 2021",issuer_name="DigiCert SHA2 Assured ID Timestamping CA",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-04-06 10:06:28"
datetime="2023-03-06 20:33:46",scan_id="033976ae-46cb-4c2e-a357-734353f7e09a",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\TWOD_Launcher.exe",signature_hash="sha256",serial_number="073637b724547cd847acfd28662a5e5b",thumbprint="281734d4592d1291d27190709cb510b07e22c405d5e0d6119b70e73589f98acf",subject_name="DigiCert Trusted G4 RSA4096 SHA256 TimeStamping CA",issuer_name="DigiCert Trusted Root G4",file_created_at="2023-03-03 23:20:42",file_modified_at="2022-04-07 09:14:08"
datetime="2023-03-06 20:33:47",scan_id="033976ae-46cb-4c2e-a357-734353f7e09a",os_version="Windows",hostname="DESKTOP-S5VJGLH",ip_address="192.168.0.23",infected_file="F:\code\pythonProject\certverify\test\VBoxSup.sys",signature_hash="sha256",serial_number="2f451139512f34c8c528b90bca471f767b83c836",thumbprint="3aa166713331d894f240f0931955f123873659053c172c4b22facd5335a81346",subject_name="VirtualBox for Legacy Windows Only Timestamp Kludge 2014",issuer_name="VirtualBox for Legacy Windows Only Timestamp CA",file_created_at="2023-03-03 23:20:43",file_modified_at="2022-10-11 08:11:56"
```


CHANGES
 -*- coding: utf-8 -*-


Changes with v1.0.0-230306

*) First Release
- To extract certificate information from arbitrary files, use the -> https://github.com/password123456/py_certificate_extractor. it may help you.
- If a driver (.sys) file is signed with a dual certificate (sha1, sha256 code signing), the sha256 certificate information is not scanned correctly. 
- This seems to be caused by the deprecation of sha1 certificates, which we are still checking.(We don't think this is a problem, and if it is, we will fix it).)
- these are Antivirus driver files, legitimate (healthy) drivers on the Windows system, and driver files that are certified by Microsoft and have a dual certificate (SHA1, SHA256).


main.py
__author__ = 'https://github.com/password123456/'
__version__ = '1.0.0-20230306'

import os
import sys
import platform
import time
import requests
import hashlib
import argparse
import uuid
import netifaces
import csv
import pefile

from datetime import datetime
from asn1crypto import cms
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.x509.oid import NameOID
from cryptography.hazmat.primitives import hashes
from concurrent.futures import ProcessPoolExecutor, as_completed

# Configurations
SCAN_EXTENSIONS = ['.exe', '.sys', '.dll']

EXCLUDED_DIRECTORIES = ['Windows', 'ProgramData']
EXCLUDED_DIRECTORIES = [subject.lower() for subject in EXCLUDED_DIRECTORIES]

TRUSTED_CODESIGN_SUBJECTS = ['Microsoft Windows', 'Microsoft Corporation', 'Microsoft Code Signing PCA', 'Microsoft Windows Production PCA']
TRUSTED_CODESIGN_SUBJECTS = [subject.lower() for subject in TRUSTED_CODESIGN_SUBJECTS]

_home_path = f'{os.getcwd()}'

_engine_file = f'{_home_path}/engine.csv'
_result_logs = f'{_home_path}/output/{datetime.today().strftime("%Y%m%d")}_scan.log'


class Bcolors:
    Black = '\033[30m'
    Red = '\033[31m'
    Green = '\033[32m'
    Yellow = '\033[33m'
    Blue = '\033[34m'
    Magenta = '\033[35m'
    Cyan = '\033[36m'
    White = '\033[37m'
    Endc = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


def download_engine():
    _url = f'https://bazaar.abuse.ch/export/csv/cscb/'
    _header = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.112 Safari/537.36',
               'Connection': 'keep-alive'}
    try:
        with open(_engine_file, 'wb') as f:
            r = requests.get(_url, headers=_header, stream=True)
            download_file_length = r.headers.get('Content-Length')
            print(f'{Bcolors.Green} Downloading: {_engine_file} / {(float(download_file_length) / (1024.0 * 1024.0)):.2f} MB {Bcolors.Endc}')

            if download_file_length is None:
                f.write(r.content)
            else:
                dl = 0
                total_length = int(download_file_length)
                start = time.perf_counter()
                for data in r.iter_content(chunk_size=8092):
                    dl += len(data)
                    f.write(data)
                    done = int(100 * dl / total_length)
                    print(f'[{">" * done}{" " * (100 - done)}] {total_length}/{dl} ({done}%) - {(time.perf_counter() - start):.2f} seconds ', end='\r')

        # Check Downloaded File
        if os.path.isfile(_engine_file):
            with open(_engine_file, 'rb') as f:
                file_read = f.read()
                file_hash = hashlib.sha256(file_read).hexdigest()
                file_info = f'===> Extracted Size: {int(os.path.getsize(_engine_file)) / (1024.0 * 1024.0):.2f} MB\n===> Hash(SHA-256) : {file_hash}\n'

                print(f'\n\n{Bcolors.Green}===> Update Success: {_engine_file} {Bcolors.Endc}')
                print(f'{Bcolors.Green}{file_info}{Bcolors.Endc}')
        else:
            print(f'{Bcolors.Yellow}[-] {_engine_file} not found. {Bcolors.Endc}')
            sys.exit(1)

    except Exception as e:
        print(f'{Bcolors.Yellow}- ::Exception:: Func:[{download_engine.__name__}] Line:[{sys.exc_info()[-1].tb_lineno}] [{type(e).__name__}] {e}{Bcolors.Endc}')


def raw_count(filename):
    n = 0
    mode = 'r'
    with open(filename, mode, encoding='utf-8') as f:
        for line in f:
            if not line.startswith('#'):
                n = n + 1
    return n


def csv_decomment(csvfile):
    for row in csvfile:
        raw = row.split('#')[0].strip()
        if raw:
            yield row


def get_engine_last_updated_date(filename):
    mode = 'r'
    with open(filename, mode, encoding='utf-8') as file:
        for line in file:
            if 'Last updated' in line:
                line = line.replace('#', '')
                line = line.lstrip().strip('\n')
                line = line.split(' ')
                line = line[2:5]
                line = ' '.join(line)
                break
    return line


def hash_exists_in_database(serial_number, thumbprint):
    mode = 'r'
    with open(_engine_file, mode, encoding='utf-8') as database:
        reader = csv.DictReader(csv_decomment(database), delimiter=',', lineterminator='\n', skipinitialspace=True)
        for row in reader:
            if str(serial_number) in str(row['serial_number']):
                if str(thumbprint) in str(row['thumbprint']):
                    return True
    return False


def scan_logs(_contents):
    output_dir = f'{_home_path}/output'
    mode = 'w'
    if os.path.exists(output_dir):
        if os.path.exists(_result_logs):
            mode = 'a'
    else:
        mode = 'w'
        os.makedirs(output_dir)
    with open(_result_logs, mode) as f:
        f.write(f'{_contents}')


def is_digitally_signed(_file_name):
    try:
        pe = pefile.PE(_file_name)
        address = pe.OPTIONAL_HEADER.DATA_DIRECTORY[pefile.DIRECTORY_ENTRY["IMAGE_DIRECTORY_ENTRY_SECURITY"]].VirtualAddress
        if address != 0:
            return True
        else:
            return False
    except pefile.PEFormatError as e:
        return False


def get_digitally_signed_info(_file_name):
    ret_signature_hash = ''
    ret_serial_number = ''
    ret_thumbprint = ''
    ret_subject_name = ''
    ret_issuer_name = ''
    is_sha256_verified = False

    try:
        pe = pefile.PE(_file_name)
        if hex(pe.DOS_HEADER.e_magic) == '0x5a4d':
            address = pe.OPTIONAL_HEADER.DATA_DIRECTORY[pefile.DIRECTORY_ENTRY["IMAGE_DIRECTORY_ENTRY_SECURITY"]].VirtualAddress
            size = pe.OPTIONAL_HEADER.DATA_DIRECTORY[pefile.DIRECTORY_ENTRY["IMAGE_DIRECTORY_ENTRY_SECURITY"]].Size

            if size > 0:
                with open(_file_name, 'rb') as fh:
                    fh.seek(address)
                    thesig = fh.read(size)

                signature = cms.ContentInfo.load(thesig[8:])
                for cert in signature['content']['certificates']:
                    x509_pem_cert = x509.load_der_x509_certificate(cert.dump(), default_backend())
                    subject_name = x509_pem_cert.subject.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value.lower()
                    if not any(subject_name.startswith(x) for x in TRUSTED_CODESIGN_SUBJECTS):
                        if x509_pem_cert.signature_hash_algorithm.name.lower() == 'sha256':
                            is_sha256_verified = True
                            ret_subject_name = x509_pem_cert.subject.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value
                            ret_issuer_name = x509_pem_cert.issuer.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value
                            ret_signature_hash = x509_pem_cert.signature_hash_algorithm.name
                            ret_thumbprint = x509_pem_cert.fingerprint(hashes.SHA256()).hex()
                            ret_serial_number = hex(x509_pem_cert.serial_number)[2:].zfill(32)

                            return True, ret_signature_hash, ret_serial_number, ret_thumbprint, ret_subject_name, ret_issuer_name

                if not is_sha256_verified:
                    return False, ret_signature_hash, ret_serial_number, ret_thumbprint, ret_subject_name, ret_issuer_name
    except (pefile.PEFormatError, AttributeError):
        return False, ret_signature_hash, ret_serial_number, ret_thumbprint, ret_subject_name, ret_issuer_name


def has_valid_extension(_file_name):
    if _file_name.endswith(tuple(SCAN_EXTENSIONS)):
        return True
    else:
        return False


def check_file_size(_file_name):
    # 10MB = '10485760'
    limit = 10485760

    f = os.stat(_file_name).st_size
    if f <= limit:
        return True
    else:
        return False


def get_creation_date(_file_name):
    if platform.system() == 'Windows':
        result = os.path.getctime(_file_name)
    else:
        result = os.path.getmtime(_file_name)
    return datetime.fromtimestamp(result).strftime('%Y-%m-%d %H:%M:%S')


def get_modified_date(_file_name):
    result = os.path.getmtime(_file_name)
    return datetime.fromtimestamp(result).strftime('%Y-%m-%d %H:%M:%S')


def get_hostname():
    return platform.node()


def get_os_version():
    return platform.system()


def get_ip_address():
    gateways = netifaces.gateways()
    default_gateway = gateways['default'][netifaces.AF_INET]
    gateway_ip, interface = default_gateway[0], default_gateway[1]
    iface = netifaces.ifaddresses(interface)
    local_ip = iface[netifaces.AF_INET][0]['addr']
    return local_ip


def create_job_id():
    return uuid.uuid4()


def check_engine():
    if os.path.exists(_engine_file):
        engine_modified_time = os.stat(_engine_file).st_mtime
        numof_ymd_today = datetime.today().strftime('%Y%m%d')
        numof_ymd_engine_file_date = datetime.fromtimestamp(engine_modified_time).strftime('%Y%m%d')

        if not(int(numof_ymd_engine_file_date) == int(numof_ymd_today)):
            get_download = False
        else:
            get_download = True

        if not get_download:
            print(f'{Bcolors.Yellow}- Updating Engine Signatures.{Bcolors.Endc}')
            download_engine()
        else:
            print(f'{Bcolors.Yellow}- Up2date Engine   : ^_^V {Bcolors.Endc}')
    else:
        print(f'{Bcolors.Yellow}- Updating Engine Signatures.{Bcolors.Endc}')
        download_engine()


def scan_file(_f_file_name):
    scan_result = ''
    if has_valid_extension(_f_file_name):
        if check_file_size(_f_file_name):
            if is_digitally_signed(_f_file_name):
                file_result, file_signature_hash, file_serial_number, file_thumbprint, file_subject_cn, file_issuer_cn = get_digitally_signed_info(_f_file_name)
                if file_result:
                    if hash_exists_in_database(file_serial_number, file_thumbprint):
                        scan_result = f'{_f_file_name}|{file_signature_hash}|{file_serial_number}|{file_thumbprint}|{file_subject_cn}|{file_issuer_cn}|{get_creation_date(_f_file_name)}|{get_modified_date(_f_file_name)}'
            return scan_result


def scan_directory(scan_path):
    log_ip_address = get_ip_address()
    log_hostname = get_hostname()
    log_scan_id = create_job_id()
    log_os_version = get_os_version()

    submitted_file_count = 0
    infected_file_count = 0
    infected_file_list = ''

    scan_start_time = time.perf_counter()

    with ProcessPoolExecutor(max_workers=5) as executor:
        for subdir, dirs, files in os.walk(scan_path):
            dirs[:] = [d for d in dirs if d not in EXCLUDED_DIRECTORIES]
            for file in files:
                try:
                    file_path = os.path.realpath(os.path.join(subdir, file))
                    result = [executor.submit(scan_file, file_path)]
                    for future in as_completed(result, timeout=10):
                        try:
                            result_str = future.result()
                            submitted_file_count += 1
                            scan_duration_time = time.perf_counter() - scan_start_time

                            if result_str:
                                infected_file_count += 1
                                infected_file_list += f' - {file_path}\n'

                                print(f'[{datetime.strftime(datetime.utcfromtimestamp(scan_duration_time), "%H:%M:%S.%f")}] '
                                      f'({submitted_file_count} scanned) {Bcolors.Blue}[infected]{Bcolors.Endc} {file_path}')

                                infected_file_info = result_str.split("|")
                                contents = (
                                    f'datetime="{datetime.today().strftime("%Y-%m-%d %H:%M:%S")}",'
                                    f'scan_id="{log_scan_id}",'
                                    f'os_version="{log_os_version}",'
                                    f'hostname="{log_hostname}",'
                                    f'ip_address="{log_ip_address}",'
                                    f'infected_file="{infected_file_info[0]}",'
                                    f'signature_hash="{infected_file_info[1]}",'
                                    f'serial_number="{infected_file_info[2]}",'
                                    f'thumbprint="{infected_file_info[3]}",'
                                    f'subject_name="{infected_file_info[4]}",'
                                    f'issuer_name="{infected_file_info[5]}",'
                                    f'file_created_at="{infected_file_info[6]}",'
                                    f'file_modified_at="{infected_file_info[7]}"\n'
                                )
                                scan_logs(contents)
                            else:
                                print(f'[{datetime.strftime(datetime.utcfromtimestamp(scan_duration_time), "%H:%M:%S.%f")}] '
                                      f'({submitted_file_count} scanned) {Bcolors.Yellow}[O.K]{Bcolors.Endc} {file_path}')
                        except TimeoutError:
                            pass
                except Exception as e:
                    # print(e)
                    continue

    if infected_file_count >= 1:
        print(f'\x1b[0;43;43m Scan Completed.! \x1b[0m \n- O.M.G... \x1b[0;42;42m [{infected_file_count}] \x1b[0m files found.')
        print(f'\n[RESULT]\n{infected_file_list}')
        print(f'>>>>> See "{_result_logs}"\n')
    else:
        _contents = f'datetime="{datetime.today().strftime("%Y-%m-%d %H:%M:%S")}",scan_id="{log_scan_id}",os="{log_os_version}",' \
                    f'hostname="{log_hostname}",ip="{log_ip_address}",infected_file="None"\n'
        scan_logs(_contents)
        print(f'\n')
        print(f'\x1b[0;43;43m Scan Completed.! \x1b[0m \n- No infected file found.! happy happy:)\n')


def main():
    print(f'\n')
    print(f'{Bcolors.Green}▌║█║▌│║▌│║▌║▌█║ {Bcolors.Red}CertVerify{Bcolors.White} v{__version__}{Bcolors.Green} ▌│║▌║▌│║║▌█║▌║█{Bcolors.Endc}\n')
    opt = argparse.ArgumentParser()
    opt.add_argument('--path', help='ex) /home/download')
    opt.add_argument('--update', action='store_true', help='Untrusted Certificates Engine Update')

    if len(sys.argv) < 1:
        opt.print_help()
        sys.exit(1)
    else:
        options = opt.parse_args()
        print(f'File scanner for files signed with leaked or untrusted certificates.\n')
        print(f'- Run time: {datetime.today().strftime("%Y-%m-%d %H:%M:%S")}')
        print('- For questions contact github.com/password123456\t\t')
        print('\n')

        if options.path:
            _scan_path = os.path.abspath(options.path)
            print(f'{Bcolors.Green}------------------------------------->{Bcolors.Endc}\n')
            check_engine()
            print(f'- Engine Updated   : {get_engine_last_updated_date(_engine_file)}')
            print(f'- Total Signatures : {raw_count(_engine_file)}')
            print(f'-{Bcolors.Green} O.K Here We go.!{Bcolors.Endc}')
            scan_directory(_scan_path)

        elif options.update:
            print(f'{Bcolors.Green}——————————————————>{Bcolors.Endc}\n')
            check_engine()
            print(f'- Engine Updated   : {get_engine_last_updated_date(_engine_file)}')
            print(f'- Total Signatures : {raw_count(_engine_file)}')
        else:
            opt.print_help()
            sys.exit(1)


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception as e:
        print(f'{Bcolors.Yellow}- ::Exception:: Func:[{__name__.__name__}] Line:[{sys.exc_info()[-1].tb_lineno}] [{type(e).__name__}] {e}{Bcolors.Endc}')
